server/rootless.go

```diff
@@ -0,0 +1,39 @@
+package server
+
+import (
+	"strings"
+
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/generate"
+)
+
+func makeOCIConfigurationRootless(g *generate.Generator) error {
+	g.Config.Linux.Resources = nil
+	g.Config.Process.OOMScoreAdj = nil
+	g.Config.Process.ApparmorProfile = ""
+	g.Config.Process.SelinuxLabel = ""
+
+	for i := range g.Config.Mounts {
+		var newOptions []string
+		for _, o := range g.Config.Mounts[i].Options {
+			if strings.HasPrefix(o, "gid=") {
+				continue
+			}
+			newOptions = append(newOptions, o)
+		}
+		g.Config.Mounts[i].Options = newOptions
+	}
+
+	g.RemoveMount("/sys")
+	sysMnt := rspec.Mount{
+		Destination: "/sys",
+		Type:        "bind",
+		Source:      "/sys",
+		Options:     []string{"nosuid", "noexec", "nodev", "ro", "rbind"},
```

Comment:
`ro, rbind` does not make `/sys/fs/cgroup` read-only recursively, could you cherry-pick https://gist.github.com/AkihiroSuda/889b29c56a3c9b546a843fb03f11d60a ?

Reply:
I am seeing an issue when I try to mount cgroup after the `/sys/` bind mount.  I am trying to fix it in podman first, as the same bind mount is used there and I am using this patch:

```diff
diff --git a/pkg/spec/spec.go b/pkg/spec/spec.go
index d9888e99..1bcb92ad 100644
--- a/pkg/spec/spec.go
+++ b/pkg/spec/spec.go
@@ -44,6 +44,13 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
                        Options:     []string{"nosuid", "noexec", "nodev", "ro", "rbind"},
                }
                g.AddMount(sysMnt)
+               cgroupMnt := spec.Mount{
+                       Destination: "/sys/fs/cgroup",
+                       Type:        "cgroup",
+                       Source:      "cgroup",
+                       Options:     []string{"nosuid", "noexec", "nodev", "relatime", "ro"},
+               }
+               g.AddMount(cgroupMnt)
        }
        if rootless.IsRootless() {
                g.RemoveMount("/dev/pts")
```

The error I see is:
```
container_linux.go:336: starting container process caused "process_linux.go:402: container init caused \"rootfs_linux.go:57: mounting \\\"/sys/fs/cgroup\\\" to rootfs \\\"/tmp/test/rootfs\\\" at \\\"/sys/fs/cgroup\\\" caused \\\"stat /tmp/test/rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/foo: no such file or directory\\\"\""
```

Reply:
Also, is it possible to exploit it without privileged helpers?

Reply:
Sorry you need to compile runc with patched `libcontainer/rootfs_linux.go`. I'll open PR in a few days.

As the issue won't be exploited without privileged helpers (e.g. pam_cgfs or manual chown invocation), the fix can be postponed.